Running test on test_events_endpoint with other test using setUpFileServer produces error

Tested on the most updated `devel` branch.

I keep getting an error about `AssertionError: The reactor had readers/writers left` in some testcases. It seems the problem relies whenever I run **together** test : `test_events_endpoint` with other test that use `setUpFileServer` function. 

It is tested with `Test/test_libtorrent_download.py` and `Test/API/test_download.py`.

--- a/Tribler/Core/TorrentDef.py
+++ b/Tribler/Core/TorrentDef.py
@@ -141,6 +141,7 @@ class TorrentDef(object):
         """
         # Class method, no locking required
         try:
+            # TODO Martijn: this request should be done using Twisted
             response = requests.get(url, timeout=30, verify=False)
             if response.ok:
                 return TorrentDef.load_from_memory(response.content)

--- a/Tribler/Core/Upgrade/pickle_converter.py
+++ b/Tribler/Core/Upgrade/pickle_converter.py
@@ -133,7 +133,7 @@ class PickleConverter(object):
             dlconfig = pickle.load(old_file)
 
         # Upgrade to new config
-        ddsconfig = DefaultDownloadStartupConfig()
+        ddsconfig = DefaultDownloadStartupConfig.getInstance()
         for key, value in dlconfig.iteritems():
             if key in ['saveas', 'max_upload_rate', 'max_download_rate', 'super_seeder', 'mode', 'selected_files',
                        'correctedfilename']:

--- a/Tribler/Test/Core/Modules/Channel/test_channel_rss.py
+++ b/Tribler/Test/Core/Modules/Channel/test_channel_rss.py
@@ -11,26 +11,35 @@ class TestChannelRss(BaseTestChannel):
 
     def setUp(self, annotate=True):
         """
-        Setup the tests by creating the ChannelRssParser instance.
+        Setup the tests by creating the ChannelRssParser instance and initializing it.
         """
         super(TestChannelRss, self).setUp(annotate=annotate)
         self.channel_rss = ChannelRssParser(self.fake_session, self.fake_channel_community, 'a')
+        self.channel_rss.initialize()
+        self.should_shutdown_after_test = True
+
+    def tearDown(self, annotate=True):
+        super(TestChannelRss, self).tearDown(annotate=annotate)
+
+        if self.should_shutdown_after_test:
+            self.channel_rss.shutdown()
 
     def test_task_scrape_no_stop(self):
+        self.channel_rss.cancel_all_pending_tasks()
         self.channel_rss._task_scrape()
         self.assertTrue(self.channel_rss.is_pending_task_active("rss_scrape"))
 
     def test_task_scrape_stop(self):
+        self.channel_rss.cancel_all_pending_tasks()
         self.channel_rss._to_stop = True
         self.channel_rss._task_scrape()
         self.assertFalse(self.channel_rss.is_pending_task_active("rss_scrape"))
 
     def test_initialize(self):
-        self.channel_rss.initialize()
         self.assertTrue(self.channel_rss.is_pending_task_active("rss_scrape"))
 
     def test_shutdown(self):
-        self.channel_rss.initialize()
+        self.should_shutdown_after_test = False
         cache_path = self.channel_rss._url_cache._file_path
         self.channel_rss._url_cache.add('a')
         self.channel_rss.shutdown()

--- a/Tribler/Test/Core/Modules/RestApi/test_events_endpoint.py
+++ b/Tribler/Test/Core/Modules/RestApi/test_events_endpoint.py
@@ -15,16 +15,18 @@ class EventDataProtocol(Protocol):
     """
     This class is responsible for reading the data received over the event socket.
     """
-    def __init__(self, messages_to_wait_for, finished):
+    def __init__(self, messages_to_wait_for, finished, response):
         self.json_buffer = []
         self.messages_to_wait_for = messages_to_wait_for + 1  # The first event message is always events_start
         self.finished = finished
+        self.response = response
 
     def dataReceived(self, data):
         self.json_buffer.append(json.loads(data))
         self.messages_to_wait_for -= 1
         if self.messages_to_wait_for == 0:
             self.finished.callback(self.json_buffer[1:])
+            self.response.connectionLost(self)
 
 
 class TestEventsEndpoint(AbstractApiTest):
@@ -34,7 +36,7 @@ class TestEventsEndpoint(AbstractApiTest):
         self.events_deferred = Deferred()
 
     def on_event_socket_opened(self, response):
-        response.deliverBody(EventDataProtocol(self.messages_to_wait_for, self.events_deferred))
+        response.deliverBody(EventDataProtocol(self.messages_to_wait_for, self.events_deferred, response))
 
     def open_events_socket(self):
         agent = Agent(reactor)

--- a/Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py
+++ b/Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py
@@ -1,14 +1,20 @@
 import json
 import base64
+import os
 import urllib
+
+import shutil
 
 from Tribler.Core.Modules.channel.channel import ChannelObject
 from Tribler.Core.Modules.channel.channel_manager import ChannelManager
+from Tribler.Core.Modules.restapi.my_channel_endpoint import MyChannelModifyTorrentsEndpoint
+from Tribler.Core.Utilities.network_utils import get_random_port
 from Tribler.Core.Utilities.twisted_thread import deferred
 from Tribler.Core.simpledefs import NTFY_CHANNELCAST
 from Tribler.Core.exceptions import DuplicateChannelNameError
 from Tribler.Test.Core.Modules.RestApi.base_api_test import AbstractApiTest
 from Tribler.Core.TorrentDef import TorrentDef
+from Tribler.Test.Core.base_test import MockObject
 from Tribler.Test.test_libtorrent_download import TORRENT_FILE
 
 
@@ -327,29 +333,35 @@ class TestMyChannelModifyTorrentsEndpoint(AbstractTestMyChannelEndpoints):
         super(TestMyChannelModifyTorrentsEndpoint, self).setUpPreSession()
         self.config.set_libtorrent(True)
 
-    @deferred(timeout=10)
     def test_add_torrent_from_url_to_my_channel_with_description(self):
+        """
+        Testing whether a torrent can be added to a channel using the API
+        """
         my_channel_id = self.create_fake_channel("channel", "")
-        torrent_path = TORRENT_FILE
 
-        @staticmethod
-        def fake_load_from_url(url):
-            return TorrentDef.load(torrent_path)
-
-        TorrentDef.load_from_url = fake_load_from_url
+        # Setup file server to serve torrent file
+        files_path = os.path.join(self.session_base_dir, 'http_torrent_files')
+        os.mkdir(files_path)
+        shutil.copyfile(TORRENT_FILE, os.path.join(files_path, 'ubuntu.torrent'))
+        file_server_port = get_random_port()
+        self.setUpFileServer(file_server_port, files_path)
 
         def verify_method_invocation(channel_id, torrent_def, extra_info={}, forward=True):
             self.assertEqual(my_channel_id, channel_id)
-            self.assertEqual(TorrentDef.load(torrent_path), torrent_def)
-            self.assertEqual({"description": "test add torrent"}, extra_info)
+            self.assertEqual(TorrentDef.load(os.path.join(files_path, 'ubuntu.torrent')), torrent_def)
+            self.assertEqual({}, extra_info)
             self.assertEqual(True, forward)
 
         self.session.add_torrent_def_to_channel = verify_method_invocation
 
-        torrent_url = 'https://www.tribler.org'
-        url = 'mychannel/torrents/' + urllib.quote_plus(torrent_url)
-        return self.do_request(url, expected_code=200, expected_json={"added": torrent_url}, request_type='PUT',
-                               post_data={"description": "test add torrent"})
+        # Martijn: We are directly invoking the endpoint here. This is because the requests library used by
+        # TorrentDef.load_from_url is not working well when performing a request using do_request and seems to hang.
+        torrent_url = 'http://localhost:%s/ubuntu.torrent' % file_server_port
+        endpoint = MyChannelModifyTorrentsEndpoint(self.session, torrent_url)
+        fake_request = MockObject()
+        fake_request.content = MockObject()
+        fake_request.content.read = lambda: ""
+        self.assertDictEqual(json.loads(endpoint.render_PUT(fake_request)), {"added": torrent_url})
 
     @deferred(timeout=10)
     def test_add_torrent_from_magnet_to_my_channel_without_description(self):
@@ -385,11 +397,10 @@ class TestMyChannelModifyTorrentsEndpoint(AbstractTestMyChannelEndpoints):
         """
         self.create_fake_channel("channel", "")
 
-        @staticmethod
-        def fake_load_from_url(url):
-            raise ValueError("Test error")
+        def fake_get_metainfo(infohash_or_magnet, callback):
+            raise ValueError(u"Test error")
 
-        TorrentDef.load_from_url = fake_load_from_url
+        self.session.lm.ltmgr.get_metainfo = fake_get_metainfo
 
         def verify_error_message(body):
             error_response = json.loads(body)
@@ -402,7 +413,7 @@ class TestMyChannelModifyTorrentsEndpoint(AbstractTestMyChannelEndpoints):
             }
             self.assertDictContainsSubset(expected_response[u"error"], error_response[u"error"])
 
-        torrent_url = 'https://www.tribler.org'
+        torrent_url = 'magnet:testtest'
         url = 'mychannel/torrents/' + urllib.quote_plus(torrent_url)
         self.should_check_equality = False
         return self.do_request(url, expected_code=500, expected_json=None, request_type='PUT')\

